planner/core/plan_stats_test.go

```diff
@@ -0,0 +1,229 @@
+			check: func(p plannercore.Plan, tableInfo *model.TableInfo) {
+				pp, ok := p.(*plannercore.PhysicalProjection)
+				c.Check(ok, IsTrue)
+				pa, ok := pp.Children()[0].(*plannercore.PhysicalApply)
+				c.Check(ok, IsTrue)
+				left := pa.PhysicalHashJoin.Children()[0]
+				right := pa.PhysicalHashJoin.Children()[0]
+				c.Assert(countFullStats(left.Stats().HistColl, tableInfo.Columns[2].ID), Greater, 0)
+				c.Assert(countFullStats(right.Stats().HistColl, tableInfo.Columns[3].ID), Greater, 0)
+			},
+		},
+		{ // CTE
+			sql: "with cte(x, y) as (select d + 1, b from t where c > 1) select * from cte where x < 3",
+			check: func(p plannercore.Plan, tableInfo *model.TableInfo) {
+				ps, ok := p.(*plannercore.PhysicalSelection)
+				c.Check(ok, IsTrue)
+				pc, ok := ps.Children()[0].(*plannercore.PhysicalCTE)
+				c.Check(ok, IsTrue)
+				pp, ok := pc.SeedPlan.(*plannercore.PhysicalProjection)
+				c.Check(ok, IsTrue)
+				reader, ok := pp.Children()[0].(*plannercore.PhysicalTableReader)
+				c.Check(ok, IsTrue)
+				c.Assert(countFullStats(reader.Stats().HistColl, tableInfo.Columns[2].ID), Greater, 0)
+			},
+		},
+	}
+	for _, testCase := range testCases {
+		if testCase.skip {
+			continue
+		}
+		is := dom.InfoSchema()
+		dom.StatsHandle().Clear() // clear statsCache
+		c.Assert(dom.StatsHandle().Update(is), IsNil)
+		stmt, err := s.ParseOneStmt(testCase.sql, "", "")
+		c.Check(err, IsNil)
+		err = executor.ResetContextOfStmt(ctx, stmt)
+		c.Assert(err, IsNil)
+		p, _, err := planner.Optimize(context.TODO(), ctx, stmt, is)
+		c.Check(err, IsNil)
+		tbl, err := is.TableByName(model.NewCIStr("test"), model.NewCIStr("t"))
+		c.Assert(err, IsNil)
+		tableInfo := tbl.Meta()
+		testCase.check(p, tableInfo)
+	}
+}
+
+func countFullStats(stats *statistics.HistColl, colID int64) int {
+	for _, col := range stats.Columns {
+		if col.Info.ID == colID {
+			return col.Histogram.Len() + col.TopN.Num()
+		}
+	}
+	return -1
+}
+
+func (s *testPlanStatsSuite) TestPlanStatsLoadTimeout(c *C) {
+	originConfig := config.GetGlobalConfig()
+	newConfig := config.NewConfig()
+	newConfig.Performance.StatsLoadConcurrency = 0
+	newConfig.Performance.StatsLoadQueueSize = 1
+	config.StoreGlobalConfig(newConfig)
+	defer config.StoreGlobalConfig(originConfig)
+	defer testleak.AfterTest(c)()
+	store, dom, err := newStoreWithBootstrap()
+	c.Check(err, IsNil)
+	defer func() {
+		dom.Close()
+		store.Close()
+	}()
+	tk := testkit.NewTestKit(c, store)
+	tk.MustExec("use test")
+	originalVal1 := tk.MustQuery("select @@tidb_stats_load_pseudo_timeout").Rows()[0][0].(string)
+	defer func() {
+		tk.MustExec(fmt.Sprintf("set global tidb_stats_load_pseudo_timeout = %v", originalVal1))
+	}()
+
+	ctx := tk.Se.(sessionctx.Context)
+	tk.MustExec("drop table if exists t")
+	tk.MustExec("set @@session.tidb_analyze_version=2")
+	tk.MustExec("set @@session.tidb_stats_load_sync_wait = 1")
```

Comment:
Can you add some brief comments on the tests. For example, why set a variable certain value and why some error occurs after some operation(just for critical step is enough).

Reply:
Done